src/charm.py

```diff
@@ -260,6 +269,185 @@ def _on_remove(self, event: RemoveEvent):
                     pass
             self._delete_vault_data()
 
+    def _on_new_vault_kv_client_attached(self, event: NewVaultKvClientAttachedEvent):
+        """Handler triggered when a new vault-kv client is attached."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can configure a vault-kv client, skipping")
+            return
+
+        if not self._is_peer_relation_created():
+            self.unit.status = WaitingStatus("Waiting for peer relation")
```

Comment:
I'm not sure about this handler dealing with status. I'd suggest logging and deferring instead and leave status management to config_changed. This is applicable at a bunch of places in this PR.